dll/win32/comctl32/listview.c

```diff
@@ -2766,7 +2769,42 @@ static void LISTVIEW_NextIconPosLeft(LISTVIEW_INFO *infoPtr, LPPOINT lpPos)
     infoPtr->currIconPos.y  = 0;
 }
 
-    
+
+/***
+ * DESCRIPTION:
+ * Returns the grid position closest to the already placed icon.
+ * The returned position is not offset by Origin.
+ *
+ * PARAMETER(S):
+ * [I] infoPtr : valid pointer to the listview structure
+ * [O] lpPos : will get the current icon position
+ * [I] nItem : item id to get position for
+ *
+ * RETURN:
+ * None
+ */
+static void LISTVIEW_NextIconPosSnap(LISTVIEW_INFO *infoPtr, LPPOINT lpPos, INT nItem)
+{
+    INT nListHeight = infoPtr->rcList.bottom - infoPtr->rcList.top;
+    INT nListWidth = infoPtr->rcList.right - infoPtr->rcList.left;
+    INT nMaxColumns = nListWidth / infoPtr->nItemWidth;
+    INT nMaxRows = nListHeight / infoPtr->nItemHeight;
+    POINT oldPosition;
+
+    // get the existing x and y position and then snap to the closest grid square
+    oldPosition.x = (LONG_PTR)DPA_GetPtr(infoPtr->hdpaPosX, nItem);
+    oldPosition.y = (LONG_PTR)DPA_GetPtr(infoPtr->hdpaPosY, nItem);
+
+    // FIXME: This could should deal with multiple icons in the same grid square
+    (*lpPos).x = max(0, (LONG)round((float)oldPosition.x / infoPtr->nItemWidth) * infoPtr->nItemWidth);
+    (*lpPos).y = max(0, (LONG)round((float)oldPosition.y / infoPtr->nItemHeight) * infoPtr->nItemHeight);
```

Comment:
I would suggest (in order to avoid floats as much as possible) the following, schematically:
```c
((oldPosition + (nItemSize/2)) / nItemSize) * nItemSize;
```
with some potential optimization of the formula.... But let's see also what the wine ppl say about this.

Reply:
Sure, I could take a look at a method to get the equivalent of round without using floats.  I just went for the quick approach.  Is there someone at wine I should reach out to first?  Or would you like me to start making this change?

Reply:
Sure you can write to the Wine guys, here is the mailing list: https://www.winehq.org/mailman/listinfo/wine-devel

Reply:
Another formula:
```c
oldPosition + (nItemSize>>1)-(oldPosition+(nItemSize>>1)) % nItemSize;
```

Reply:
Nice, no division!  I've updated the PR to include this suggested formula.

Reply:
Nice, no division required.  I've updated the code you suggested.  I'll reach out to Wine as well.  Will my headway with wine affect the ability for me to continue working on this in reactos?

Reply:
There is no problem sending patches to Wine as well if you find that code from them that we use in ReactOS has a bug or is incomplete. Many of the developers there already did that, and it eases code synchronizations for later.
In case you send more substantial patches targeting some Wine code they will usually ask you to write a unit test to test whether what you wrote behaves similarly than on Windows.

Reply:
Yet another formula that works ok for negative positions:
```c
newPosition = ( newPosition = oldPosition + (1|-(oldPosition<0))*(nItemSize>>1) ) - (newPosition%nItemSize);
```
(with some inspiration from http://graphics.stanford.edu/~seander/bithacks.html#CopyIntegerSign )
But I think at the end Wine might prefer using the floats with round()...